reference/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,349 @@
+        p_hat, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> v_entry{
+        v, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> s_entry{
+        s, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> s_hat_entry{
+        s_hat, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> t_entry{
+        t, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> rho_old_entry{
+        rho_old, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> rho_new_entry{
+        rho_new, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> omega_entry{
+        omega, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> alpha_entry{
+        alpha, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> temp_entry{
+        temp, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<real_type> rhs_norms_entry{
+        norms_rhs, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<real_type> res_norms_entry{
+        norms_res, num_rhs, 1, num_rhs};
+
+    // generate preconditioner
+    prec.generate(batch_item_id, A_entry, prec_work);
+
+    // initialization
+    // rho_old = 1, omega = 1, alpha = 1
+    // compute b norms
+    // r = b - A*x
+    // compute residual norms
+    // r_hat = r
+    // p = 0
+    // p_hat = 0
+    // v = 0
+    initialize(A_entry, b_entry, gko::batch::to_const(x_entry), rho_old_entry,
+               omega_entry, alpha_entry, r_entry, r_hat_entry, p_entry,
+               p_hat_entry, v_entry, rhs_norms_entry, res_norms_entry);
+
+    // stopping criterion object
+    StopType stop(settings.residual_tol, rhs_norms_entry.values);
+
+    int iter{};
+
+    for (iter = 0; iter < settings.max_iterations; iter++) {
+        if (stop.check_converged(res_norms_entry.values)) {
+            logger.log_iteration(batch_item_id, iter,
+                                 res_norms_entry.values[0]);
+            break;
+        }
+
+        // rho_new =  < r_hat , r > = (r_hat)' * (r)
+        compute_dot_product_kernel<ValueType>(gko::batch::to_const(r_hat_entry),
+                                              gko::batch::to_const(r_entry),
+                                              rho_new_entry);
+
+        // beta = (rho_new / rho_old)*(alpha / omega)
+        // p = r + beta*(p - omega * v)
+        update_p(gko::batch::to_const(rho_new_entry),
+                 gko::batch::to_const(rho_old_entry),
+                 gko::batch::to_const(alpha_entry),
+                 gko::batch::to_const(omega_entry),
+                 gko::batch::to_const(r_entry), gko::batch::to_const(v_entry),
+                 p_entry);
+
+        // p_hat = precond * p
+        prec.apply(gko::batch::to_const(p_entry), p_hat_entry);
```

Comment:
inspired from cuda implementation, p_hat does not need to be initialized in initialze step